dev-tools/releng/src/main.rs

```diff
@@ -0,0 +1,669 @@
+                    shell_words::quote(args.helios_dir.as_str())
+                );
+                preflight_ok = false;
+            }
+        }
+    } else {
+        info!(logger, "cloning helios to {}", args.helios_dir);
+        Command::new("git")
+            .args(["clone", "https://github.com/oxidecomputer/helios.git"])
+            .arg(&args.helios_dir)
+            .ensure_success(&logger)
+            .await?;
+    }
+    // Record the branch and commit in the output
+    Command::new("git")
+        .arg("-C")
+        .arg(&args.helios_dir)
+        .args(["status", "--branch", "--porcelain=2"])
+        .ensure_success(&logger)
+        .await?;
+
+    // Check that the omicron1 brand is installed
+    if !Command::new("pkg")
+        .args(["verify", "-q", "/system/zones/brand/omicron1/tools"])
+        .is_success(&logger)
+        .await?
+    {
+        error!(
+            logger,
+            "the omicron1 brand is not installed; install it with \
+            `pfexec pkg install /system/zones/brand/omicron1/tools`"
+        );
+        preflight_ok = false;
+    }
+
+    // Check that the datasets for helios-image to use exist
+    for (dataset, option) in [
+        (&args.host_dataset, "--host-dataset"),
+        (&args.recovery_dataset, "--recovery-dataset"),
+    ] {
+        if !Command::new("zfs")
+            .arg("list")
+            .arg(dataset)
+            .is_success(&logger)
+            .await?
+        {
+            error!(
+                logger,
+                "the dataset {0} does not exist; run `pfexec zfs create \
+                -p {0}`, or specify a different one with {1}",
+                shell_words::quote(dataset),
+                option
+            );
+            preflight_ok = false;
+        }
+    }
+
+    if !preflight_ok {
+        bail!("some preflight checks failed");
+    }
+
+    fs::create_dir_all(&args.output_dir).await?;
+
+    // DEFINE JOBS ============================================================
+    let tempdir = camino_tempfile::tempdir()
+        .context("failed to create temporary directory")?;
+    let mut jobs = Jobs::new(&logger, permits.clone(), &args.output_dir);
+
+    jobs.push_command(
+        "helios-setup",
+        Command::new("ptime")
+            .args(["-m", "gmake", "setup"])
+            .current_dir(&args.helios_dir)
+            // ?!?!
+            // somehow, the Makefile does not see a new `$(PWD)` without this.
+            .env("PWD", &args.helios_dir)
+            // Setting `BUILD_OS` to no makes setup skip repositories we don't
+            // need for building the OS itself (we are just building an image
+            // from an already-built OS).
+            .env("BUILD_OS", "no"),
+    );
+
+    // Download the toolchain for phbl before we get to the image build steps.
+    // (This is possibly a micro-optimization.)
+    jobs.push_command(
+        "phbl-toolchain",
+        Command::new("cargo")
```

Comment:
Same, consult `CARGO` first (this is important because rustup will set it).

Reply:
In this particular case (and many other cases) we explicitly _do not_ want that; we are deliberately calling the rustup proxy here to pull down [the toolchain that phbl wants](https://github.com/oxidecomputer/phbl/blob/88977630a62b27f45b6d3ccc394eb3c708f95a44/rust-toolchain). (This is why `CARGO`, `CARGO_MANIFEST_DIR`, and `RUSTUP_TOOLCHAIN` are unset early; the Helios build process uses a menagerie of toolchain versions.)

Reply:
Ohhhh yeah. Missed that. Agreed that clear variable separation would be ideal.

Reply:
So it turns out you should not trust the `CARGO` set by rustup without also keeping `RUSTUP_TOOLCHAIN` set. The resulting problem is somewhat unintuitive to debug but very funny once you figure out what's going on:

- `CARGO` is set by the rustup proxy to `/home/iliana/.rustup/toolchains/1.77.2-x86_64-unknown-illumos/bin/cargo`.
- When this is run, Cargo searches for rustc in `PATH`, and finds the rustup proxy at `~/.cargo/bin/rustc`.
- The rustup-proxied rustc is invoked to compile some dependency at some path in `~/.cargo/registry`.
- Cargo invokes rustc on a crate that ships a rust-toolchain.toml, or a Git checkout that has a rust-toolchain.toml.
- The rustc proxy, not having any override set in RUSTUP_TOOLCHAIN, proceeds to fetch and use the specified toolchain, then is extremely confused when it sees a dependency that was compiled by somebody else:

```
[ omicron-package] error[E0514]: found crate `syn` compiled by an incompatible version of rustc
[ omicron-package]   --> /home/iliana/.cargo/git/checkouts/opte-c3062ea19cee8fd4/7ee353a/crates/kstat-macro/src/lib.rs:61:13
[ omicron-package]    |
[ omicron-package] 61 |             syn::Fields::Unit => {
[ omicron-package]    |             ^^^
[ omicron-package]    |
[ omicron-package]    = note: the following crate versions were found:
[ omicron-package]            crate `syn` compiled by rustc 1.77.2 (25ef9e3d8 2024-04-09): /home/iliana/git/omicron/target/release/deps/libsyn-39c2edc7c7bde633.rlib
[ omicron-package]    = help: please recompile that crate using this compiler (rustc 1.76.0 (07dca489a 2024-02-04)) (consider running `cargo clean` first)
```

Overall given our dependence on using the rustup proxy to correctly build the Helios build utilities, I think we should stick to using the "cargo" on the user's path universally instead of attempting to consult `CARGO`.

Reply:
Actually, now I have a commit I'm testing locally that doesn't unset any environment at the start, and unsets environment variables when needed when running particular commands. This should allow for, say `cargo +nightly xtask releng` while still using the correct toolchains for the Helios stuff.